### PR TITLE
Adding plots to ITSTPC matching QC

### PR DESCRIPTION
**Pull request by chiarazampolli**
Added: 
- matching with ITS, 
- mathcing vs 1./pt, 
- TPC tracks DCA, 
- matching vs pt removing eta ~0.

@noferini 

To test:

o2-global-track-cluster-reader --tpc-track-reader "tpctracks.root --reader-delay 10" --shm-segment-size 64000000000 --disable-mc --hbfutils-config o2_tfidinfo.root --track-types "ITS,TPC,ITS-TPC" --cluster-types "TPC,ITS" | o2-qc --config json:///home/zampolli/work/O2/RelVal/CPU_6/QC_GLO.json -b

Update of O2DPG json to be done.

Corresponding changes in QC: https://github.com/AliceO2Group/QualityControl/pull/2001

Comparison of matching vs pt with previous version:

Screenshot from 2023-09-27 18-12-51
Screenshot from 2023-09-27 18-12-45
Screenshot from 2023-09-27 18-12-38

### CODE REVIEW

**Comment by chiarazampolli** [comment]
Moving to WIP while I fix the formatting etc.

**Comment by chiarazampolli** [comment]
Hello @shahor02 , @noferini , @JianLIUhep, @knopers8 , @Barthelemy ,
Again we have interdependecies bwtween QC and O2. 
I think we should merge this one, if O2 compiled (QC will not for sure). Then QC will be broken, and we'll need to merge https://github.com/AliceO2Group/QualityControl/pull/2001.
Then I will update the json for sync and async in O2DPG.
Cheers,
Chiara

**Comment by shahor02** [comment]
@chiarazampolli if you keep unchanged old getters (even ie making them dummy when needed) and add new ones, the QC should not fail. Then, after merging this PR, then QC one, one can suppress old methods.

**Comment by chiarazampolli** [comment]
Hello @shahor02 , 
Since I now removed everything that was old, isn't it enough to see that the O2 compilation works, and QC breaks where it is expected to break?
Cheers,
Chiara

**Comment by chiarazampolli** [comment]
Uploading json file to test. In order to upload, I had to change the extension, but it is a json file.
[QC_GLO.txt](https://github.com/AliceO2Group/AliceO2/files/12754222/QC_GLO.txt)

**Comment by shahor02** [comment]
Then the QC will be broken for a day or so. Could just add the lines below somewhere before `private`?
```
  // to remove after merging QC PR
  TH1D* getHistoPt() const { return nullptr; }                               // old
  TH1D* getHistoPtTPC() const { return nullptr; }                            // old
  TEfficiency* getFractionITSTPCmatch() const { return nullptr; }            // old
  TH1F* getHistoPhi() const { return nullptr; }                              // old
  TH1F* getHistoPhiTPC() const { return nullptr; }                           // old
  TEfficiency* getFractionITSTPCmatchPhi() const { return nullptr; }         // old
  TH2F* getHistoPhiVsPt() const { return nullptr; }                          // old
  TEfficiency* getFractionITSTPCmatchPhiVsPt() const { return nullptr; }     // old
  TH1F* getHistoEta() const { return nullptr; }                              // old
  TEfficiency* getFractionITSTPCmatchEta() const { return nullptr; }         // old
  TH2F* getHistoEtaVsPt() const { return nullptr; }                          // old
  TH2F* getHistoEtaVsPtTPC() const { return nullptr; }                       // old
  TEfficiency* getFractionITSTPCmatchEtaVsPt() const { return nullptr; }     // old
  TH1F* getHistoPtPhysPrim() const { return nullptr; }                       // old
  TH1F* getHistoPtTPCPhysPrim() const { return nullptr; }                    // old
  TEfficiency* getFractionITSTPCmatchPhysPrim() const { return nullptr; }    // old
  TH1F* getHistoPhiPhysPrim() const { return nullptr; }                      // old
  TEfficiency* getFractionITSTPCmatchPhiPhysPrim() const { return nullptr; } // old
  TH1F* getHistoEtaPhysPrim() const { return nullptr; }                      // old
  TH1F* getHistoEtaTPCPhysPrim() const { return nullptr; }                   // old
  TEfficiency* getFractionITSTPCmatchEtaPhysPrim() const { return nullptr; } // old
```

**Comment by chiarazampolli** [comment]
Hello @shahor02 , 
Sure, I will change it. 
Cheers,
Chiara

**Comment by chiarazampolli** [comment]
Hello @shahor02 , 
I have the impression that now the failures are unrelated. Could you please check?
Cheers,
Chiara

**Comment by shahor02** [comment]
Actually, the CIs fail to start with `Git fetch failed`.
@TimoWilken could you please check?

**Comment by chiarazampolli** [comment]
Hello @shahor02 , @TimoWilken , 
I don't think the errors are related to the PR. Can it be reviewed?
Cheers,
Chiara

**Comment by shahor02** [comment]
Hi @chiarazampolli 

the macOS-arm errors are genuine: to compose histo title you use more printf format specifiers than arguments:
```
/Volumes/build/alice-ci-workdir/o2/sw/SOURCES/O2/11974/0/Detectors/GlobalTracking/src/MatchITSTPCQC.cxx:149:105: error: more '%' conversions than data arguments [-Werror,-Wformat-insufficient-args]
    mPhiDen[i] = new TH1F(Form("mPhiDen_%s", title[i].c_str()), Form("Phi distribution of %s tracks %s %s; Phi [rad] %s; dNdPhi", title[i].c_str(), etaSel[i].c_str()), 100, 0.f, 2 * TMath::Pi());
   ...
```
Besides that the fullCI seg.faults in the qc-task-GLO-MTCITSTPC: I guess this is because current QC calls the dummy getters we put in this PR to not break the QC compilation. If in you local tests with modified QC it works fine, we can ignore the fullCI error.

**Comment by chiarazampolli** [comment]
Hello @shahor02 , 
I saw the failure in the CI, and did not understand it was a crash. Locally, with updated QC, it works. Unfortunately, when two packages depend on each other, I think some failures are to be expected.
I will fix mac, sure. 
Thanks for taking a look.
Chiara

**Comment by chiarazampolli** [comment]
Hello @shahor02 , 
I would say that now it is fine, right?
Chiara

**Comment by martenole** [comment]
But isn't there a way to avoid breaking the fullCI?
```
[47960:internal-dpl-ccdb-backend]: [05:36:27][INFO] ccdb reads http://alice-ccdb.cern.ch/MCH/Calib/CTFDictionaryTree/1546300800000/e4eb37f0-06d8-11ee-b4d8-200114580201 for 1550600800000 (load to memory, agent_id: alimetal04.cern.ch-1696311382-bI3HMd), 
[48073:qc-task-GLO-MTCITSTPC]: *** Program crashed (Segmentation fault, FPE, BUS, ABRT, KILL, Unhandled Exception, ...)
[48073:qc-task-GLO-MTCITSTPC]: Backtrace by DPL:
[48073:qc-task-GLO-MTCITSTPC]: SEGMENTATION FAULT
[48073:qc-task-GLO-MTCITSTPC]: Executable is /sw/slc8_x86-64/QualityControl/v1.118.0-local48/bin/o2-qc
[47960:internal-dpl-ccdb-backend]: Info in <TJAlienFile::Open>: Accessing file /alice/data/CCDB/MFT/Calib/CTFDictionaryTree/05/53623/a61db061-06d8-11ee-b4d8-200114580201 in SE <ALICE::CERN::OCDB>
[48073:qc-task-GLO-MTCITSTPC]:     /lib64/libc.so.6:     ?? ??:0
[48073:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/QualityControl/v1.118.0-local48/lib/libO2QualityControl.so: o2::quality_control::core::ObjectsManager::startPublishing(TObject*)
[48073:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/QualityControl/v1.118.0-local48/lib/libO2QcGLO.so: o2::quality_control_modules::glo::ITSTPCMatchingTask::initialize(o2::framework::InitContext&)
[48073:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/QualityControl/v1.118.0-local48/lib/libO2QualityControl.so: o2::quality_control::core::TaskRunner::init(o2::framework::InitContext&)
[48073:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/QualityControl/v1.118.0-local48/lib/libO2QualityControl.so: o2::framework::adaptFromTask<o2::quality_control::core::TaskRunner, o2::quality_control::core::TaskRunner>(o2::quality_control::core::TaskRunner&&)::{lambda(o2::framework::InitContext&)#1}::operator()(o2::framework::InitContext&) const
[48073:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/QualityControl/v1.118.0-local48/lib/libO2QualityControl.so: std::_Function_handler<std::function<void (o2::framework::ProcessingContext&)> (o2::framework::InitContext&), o2::framework::adaptFromTask<o2::quality_control::core::TaskRunner, o2::quality_control::core::TaskRunner>(o2::quality_control::core::TaskRunner&&)::{lambda(o2::framework::InitContext&)#1}>::_M_invoke(std::_Any_data const&, o2::framework::InitContext&)
[48073:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/O2/11974-slc8_x86-64-local2/lib/libO2Framework.so: o2::framework::DataProcessingDevice::Init()
[48073:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/FairMQ/v1.8.0-6/lib/libfairmq.so.1.8.0: fair::mq::Device::InitWrapper()
[48073:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/FairMQ/v1.8.0-6/lib/libfairmq.so.1.8.0: boost::detail::function::void_function_obj_invoker1<std::function<void (fair::mq::State)>, void, fair::mq::State>::invoke(boost::detail::function::function_buffer&, fair::mq::State)
[48073:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/FairMQ/v1.8.0-6/lib/libfairmq.so.1.8.0: boost::signals2::detail::signal_impl<void (fair::mq::State), boost::signals2::optional_last_value<void>, int, std::less<int>, boost::function<void (fair::mq::State)>, boost::function<void (boost::signals2::connection const&, fair::mq::State)>, boost::signals2::mutex>::operator()(fair::mq::State)
[48073:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/FairMQ/v1.8.0-6/lib/libfairmq.so.1.8.0: fair::mq::fsm::Machine_::ProcessWork()
[48073:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/FairMQ/v1.8.0-6/lib/libfairmq.so.1.8.0: fair::mq::StateMachine::ProcessWork()
[48073:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/FairMQ/v1.8.0-6/lib/libfairmq.so.1.8.0: fair::mq::DeviceRunner::Run()
[48073:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/O2/11974-slc8_x86-64-local2/lib/libO2Framework.so: doChild(int, char**, o2::framework::ServiceRegistry&, o2::framework::RunningWorkflowInfo const&, o2::framework::RunningDeviceRef, o2::framework::DriverConfig const&, o2::framework::ProcessingPolicies, std::__cxx11::basic_string<char, 
```
E.g. changing QC first and add a protection or simply disabling the task in the O2DPG json file in O2DPG used for the fullCI?

**Comment by chiarazampolli** [comment]
Hello @martenole , 
Either O2 or QC will be broken, I don't know how to overcome it. QC is now waiting for this PR.
Cheers,
Chiara

**Comment by shahor02** [comment]
This would require making old (obsolete) getters (used by QC) aliased to new getters instead of returning nullptr (which makes to QC crash). I don't think it is worth it: the fullCI fails where it is expected to fail in the QC and locally, with fixed QC Chiara can run fullCI (@chiarazampolli  can you confirm that you run fullCI and not the FST only)
```
WORKFLOW_EXTRA_PROCESSING_STEPS=TPC_DEDX,MFT_RECO,MID_RECO,MCH_RECO,MATCH_MFTMCH,MATCH_MCHMID,MUON_SYNC_RECO,ZDC_RECO FST_SYNC_EXTRA_WORKFLOW_PARAMETERS=QC,CALIB_LOCAL_AGGREGATOR,CALIB_LOCAL_INTEGRATED_AGGREGATOR QC_REDIRECT_MERGER_TO_LOCALHOST=1 GEN_TOPO_WORKDIR=`pwd` ALICE_O2SIM_DUMPLOG=1 NEvents=5 NEventsQED=100 O2SIMSEED=12345 $O2_ROOT/prodtests/full_system_test.sh
$O2_ROOT/prodtests/full_system_test_ci_extra_tests.sh
```

**Comment by chiarazampolli** [comment]
Hi @shahor02 , 
The command I ran is:
```
o2-global-track-cluster-reader --tpc-track-reader "tpctracks.root --reader-delay 10" --shm-segment-size 64000000000 --disable-mc --hbfutils-config o2_tfidinfo.root --track-types "ITS,TPC,ITS-TPC" --cluster-types "TPC,ITS" | o2-qc --config json:///home/zampolli/work/O2/RelVal/CPU_6/QC_GLO.json -b
```
as reported at the beginning of the PR. I can also test the fullCI.
Chiara

**Comment by chiarazampolli** [comment]
Hello, 

fullCI needs an update in the json in O2DPG, as I planned to do after this PR. At this point, I will do it now. 
But the command:

{code}
$O2_ROOT/prodtests/full_system_test_ci_extra_tests.s
{code}

did not work for me, something unrelated, I would say:

{code}
[O2PDPSuite/latest-o2] ~/work/O2/fullCI $> more o2-raw-file-reader-workflow.log
Running: TIME="#walltime %e" /usr/bin/time --output=o2-raw-file-reader-workflow.log_time ./o2-raw-file-reader-workflow.log_tmp.sh
[INFO] WorkflowSpec defineDataProcessing
[INFO] WorkflowSpec FLPWorkflow
[INFO] ft0-datareader-dpl
[ERROR] error parsing options of o2-ft0-flp-dpl-workflow: the required argument for option '--shm-segment-size' is missing
TASK-EXIT-CODE: 1
{code}

I am now opening the PR in O2DPG.

Chiara

**Comment by chiarazampolli** [comment]
Here it is:

https://github.com/AliceO2Group/O2DPG/pull/1257/files

@shahor02 , if you agree, we can merge it.

**Comment by chiarazampolli** [comment]
Hello @shahor02 , 
The CI re-run, but indeed it is failing in publishing. 
```
[47568:qc-task-GLO-MTCITSTPC]:     /lib64/libc.so.6:     ?? ??:0
[47568:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/QualityControl/v1.118.0-local65/lib/libO2QualityControl.so: o2::quality_control::core::ObjectsManager::startPublishing(TObject*)
[47568:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/QualityControl/v1.118.0-local65/lib/libO2QcGLO.so: o2::quality_control_modules::glo::ITSTPCMatchingTask::initialize(o2::framework::InitContext&)
[47568:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/QualityControl/v1.118.0-local65/lib/libO2QualityControl.so: o2::quality_control::core::TaskRunner::init(o2::framework::InitContext&)
[47568:qc-task-GLO-MTCITSTPC]:     /sw/slc8_x86-64/QualityControl/v1.118.0-local65/lib/libO2QualityControl.so: o2::framework::adaptFromTask<o2::quality_control::core::TaskRunner, o2::quality_control::core::TaskRunner>(o2::quality_control::core::TaskRunner&&)::{lambda(o2::framework::InitContext&)#1}::operator()(o2::framework::InitContext&) const
```
Since the second test you proposed failed, but not due to QC, maybe we can anyway merge? Have you reviewed the PR?
Cheers,
Chiara

**Comment by shahor02** [comment]
Hi @chiarazampolli ,

Sorry, I don't understand, why it is failing in publishing if you are using locally fixed QC?
I've checked your PR and haven't seen obvious problems.

Cheers,
 Ruben

**Comment by chiarazampolli** [comment]
Hello @shahor02 , 

The publishing is failing in the fullCI not in my local test. The log in https://github.com/AliceO2Group/AliceO2/pull/11974#issuecomment-1746723580, is from fullCI.

My local test fails in the second test you proposed, see https://github.com/AliceO2Group/AliceO2/pull/11974#issuecomment-1745118475
```
[O2PDPSuite/latest-o2] ~/work/O2/fullCI $> more o2-raw-file-reader-workflow.log
Running: TIME="#walltime %e" /usr/bin/time --output=o2-raw-file-reader-workflow.log_time ./o2-raw-file-reader-workflow.log_tmp.sh
[INFO] WorkflowSpec defineDataProcessing
[INFO] WorkflowSpec FLPWorkflow
[INFO] ft0-datareader-dpl
[ERROR] error parsing options of o2-ft0-flp-dpl-workflow: the required argument for option '--shm-segment-size' is missing
TASK-EXIT-CODE: 1
```

Cheers,

Chiara

**Comment by shahor02** [comment]
I understand that it is the fullCI, this was the point: to run locally fullCI with local QC patch applied. If this was the case, why does it fail?
The ull_system_test_ci_extra_tests.sh is irrelevant, this is FT0 raw data reading test.

**Comment by chiarazampolli** [comment]
I meant it is fullCI from the CI tests here.

**Comment by chiarazampolli** [comment]
Hello @shahor02 , 

I am sorry to bother again, I would like to understand how we should proceed here. I repeat the situation:

- local fullCI works;
- fullCI from the CI is broken due to not updated QC.

Cheers,

Chiara

**Comment by shahor02** [comment]
Hi @chiarazampolli 
then I misunderstood you: if you confirm that locally (with fixed QC) the fullCI  with QC (with the command line as in https://github.com/AliceO2Group/AliceO2/pull/11974#issuecomment-1744810919) works then I merge this PR.

**Comment by chiarazampolli** [comment]
Yes, it does.

cheers,

chiara

**Review by shahor02 — APPROVED** [review]

**Pull request merged**
